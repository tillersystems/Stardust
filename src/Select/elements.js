import styled from 'styled-components';

export const HeaderButton = styled.button`
  display: flex;
  align-items: center;
  height: 100%;
  width: 100%;
  position: relative;
  padding: 0 2.4rem 0 1.6rem;
  cursor: pointer;
  text-align: left;
  font-size: ${({ theme: { fonts } }) => fonts.size.medium};
  border: 1px solid ${({ theme: { palette } }) => palette.lightGrey};
  border-radius: ${({ theme: { dimensions } }) => dimensions.radius};
  background: linear-gradient(
    180deg,
    ${({ theme: { palette } }) => palette.white} 0%,
    ${({ theme: { palette } }) => palette.paleGrey} 100%
  );
  color: ${({ theme: { palette } }) => palette.spaceGrey};

  &::after {
    content: '';

    position: absolute;
    right: 1.4rem;
    top: 37%;

    width: 0;
    height: 0;

    border: 0 solid transparent;
    border-right-width: 3px;
    border-left-width: 3px;
    border-bottom: 3px solid ${({ theme: { palette } }) => palette.spaceGrey};
  }

  &::before {
    content: '';

    position: absolute;
    right: 1.4rem;
    bottom: 33%;

    width: 0;
    height: 0;

    border: 0 solid transparent;
    border-left-width: 3px;
    border-right-width: 3px;
    border-top: 3px solid ${({ theme: { palette } }) => palette.spaceGrey};
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.4;
  }
`;

export const HeaderContent = styled.div`
  vertical-align: middle;
  display: inline-block;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  max-width: 100%;
`;

export const Container = styled.div`
  position: relative;
  user-select: none;
  height: 3.8rem;
`;
