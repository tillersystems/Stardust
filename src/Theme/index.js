/**
 * Theme exported to the provider
 * This is globally available in styled-components when interpolating a function like so:
 * ${(props) => props.theme}
 * Or using import { withTheme } from 'styled-components';
 */

import { breakpoints } from './breakpoints';
import { palette } from './palette';
import { dimensions } from './dimensions';
import { fonts } from './fonts';

const Theme = {
  breakpoints,
  palette,
  dimensions,
  fonts,
};

export default Theme;
