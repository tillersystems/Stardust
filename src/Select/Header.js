import React, { Children } from 'react';
import PropTypes from 'prop-types';

import { HeaderButton, HeaderContent } from './elements';

/**
 * SearchBar
 *
 * This component is in charge of displaying
 * a search bar
 *
 * @param {bool} disabled - If button is disabled
 * @param {function} onClick - Callback function called when button is clicked
 * @param {bool} isOpen - When popover is open
 * @param {string} placeholder - Placeholder of button when no value is selected
 * @param {node} children - The content of Header
 *
 * @return {jsx}
 */
const Header = ({ disabled, onClick, isOpen, placeholder, children }) => {
  // Remove empty children to ensure placeholder is display if empty
  const filteredChildren = Children.toArray(children).filter(Boolean);

  return (
    <HeaderButton
      type="button"
      disabled={disabled}
      onClick={onClick}
      aria-haspopup="true"
      aria-expanded={isOpen}
      data-testid="select-button"
    >
      <HeaderContent>{filteredChildren.length ? filteredChildren : placeholder}</HeaderContent>
    </HeaderButton>
  );
};

const { bool, func, node, string } = PropTypes;

/** Prop types. */
Header.propTypes = {
  children: node,
  disabled: bool,
  isOpen: bool,
  onClick: func.isRequired,
  placeholder: string,
};

/** Default props. */
Header.defaultProps = {
  children: null,
  disabled: false,
  isOpen: false,
  placeholder: '',
};

export default Header;
