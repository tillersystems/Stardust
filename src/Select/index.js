import React, { Children, PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import posed, { PoseGroup } from 'react-pose';

import { Header, Menu, MenuItem } from './elements';
import Option from './Option';

const { bool, func, node, string } = PropTypes;

/**
 * Select
 *
 * This component is in charge of displaying
 * a select
 *
 * @param {node} children // Anything that can be rendered: numbers, strings, elements or an array (or fragment).
 * @param {string} className /// className needed by styled components.
 * @param {function} onChange // Function fired when an element of the <Select /> is selected.
 * @param {function} onToggle // Function fired when the <Select /> is toggled.
 * @param {string} placeholder // placeholder of the <Select /> component.
 * @param {boolean} resetValue // Title of the <Select /> component.
 *
 * @return {jsx}
 */

class Select extends PureComponent {
  static Option = Option;

  /** Prop types. */
  static propTypes = {
    children: node,
    className: string,
    disabled: bool,
    onChange: func,
    onToggle: func,
    placeholder: string,
    resetValue: bool,
    // https://github.com/yannickcr/eslint-plugin-react/issues/1520
    // eslint-disable-next-line react/no-unused-prop-types
    width: string,
    initialValue: string,
  };

  /** Default props. */
  static defaultProps = {
    children: null,
    className: '',
    disabled: false,
    onChange: () => {},
    onToggle: () => {},
    resetValue: false,
    placeholder: null,
    width: '100%',
    initialValue: null,
  };

  /**
   * getDerivedStateFromProps
   *
   * @param {object} props
   * @param {object} state
   *
   * @return {object}
   */
  static getDerivedStateFromProps(props, state) {
    if (props.resetValue !== state.resetValue) {
      return {
        value: null,
      };
    }

    return null;
  }

  /** Internal state. */
  state = {
    displayMenu: false,
    placeholder: this.props.placeholder, // eslint-disable-line react/destructuring-assignment
    value: null,
    resetValue: this.props.resetValue, // eslint-disable-line react/destructuring-assignment
  };

  /**
   * Component lifecycle method
   * Once the component is mounted, checks if a value is provided
   * to the component to initialize the displayed value on the Select.
   * Otherwise, if a placeholder is provided, no value is initialized to display
   * the placeholder at the render.
   */
  componentDidMount() {
    const { children, initialValue, placeholder } = this.props;

    if (initialValue) {
      const isMatchingOption =
        children.filter(option => option.props.value === initialValue).length === 1;

      if (isMatchingOption) {
        this.setState({ value: initialValue });
      } else {
        this.initializeValue();
      }
    } else if (!placeholder) {
      this.initializeValue();
    }
  }

  /**
   * Initialize in the state the value displayed on the select
   * Takes the first option available in the children props
   *
   */
  initializeValue = () => {
    const {
      children: [
        {
          props: { value: firstValue },
        },
      ],
    } = this.props;

    this.setState({ value: firstValue });
  };

  /**
   * Toogle Menu
   */
  toggleMenu = () => {
    const { displayMenu } = this.state;
    const { onToggle } = this.props;

    this.setState(
      prevState => ({
        ...prevState,
        displayMenu: !prevState.displayMenu,
      }),
      () => {
        onToggle && onToggle(!displayMenu);
      },
    );
  };

  /**
   * Handle Selected
   *
   * @param {string} value
   *
   */
  handleSelected(value) {
    const { onChange } = this.props;

    this.setState(
      prevState => ({
        ...prevState,
        value,
      }),
      () => {
        onChange && onChange(value);
        this.toggleMenu();
      },
    );
  }

  /**
   * Renders the component.
   *
   * @return {jsx}
   */
  render() {
    const { children, className, disabled } = this.props;
    const { displayMenu, placeholder, value } = this.state;

    return (
      <div className={className}>
        <Header
          onClick={!disabled ? this.toggleMenu : null}
          disabled={disabled}
          aria-haspopup="true"
          aria-expanded={displayMenu}
        >
          {/* if placeholder is defined display it, otherwise use the value of the first option */}
          {placeholder && !value
            ? placeholder
            : Children.map(children, child => (child.props.value === value ? child : null))}
        </Header>
        <PoseGroup>
          {displayMenu && (
            <MenuAnimation key="Menu" role="menu">
              {Children.map(children, child => (
                <MenuItem
                  key={child}
                  role="menuitem"
                  onClick={() => this.handleSelected(child.props.value)}
                >
                  {child}
                </MenuItem>
              ))}
            </MenuAnimation>
          )}
        </PoseGroup>
      </div>
    );
  }
}

/**
 * Animation
 */
const MenuAnimation = posed(Menu)({
  enter: {
    opacity: 1,
    scaleY: 1,
    transition: {
      scaleY: { ease: 'anticipate', duration: 150 },
      default: { duration: 150 },
    },
  },
  exit: {
    opacity: 0,
    scaleY: 0,
    transition: { duration: 150 },
  },
});

export default styled(Select)`
  position: relative;
  user-select: none;
  width: ${({ width }) => (/^\d+(rem|px)$/.test(width) ? width : 'auto')};

  /* Disabled */
  ${({ disabled }) =>
    disabled &&
    css`
      ${Header} {
        cursor: not-allowed;
        opacity: 0.4;
      }
    `}
`;
