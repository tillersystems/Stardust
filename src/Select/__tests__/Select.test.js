import React from 'react';
import 'jest-styled-components';

import Select from '..';
import { Icon } from '../..';
import Theme from '../../Theme';

describe('<Select />', () => {
  it('should render without a problem', () => {
    const render = shallowWithTheme(
      <Select>
        <option value="home">Home</option>
        <option value="calendar">Calendar</option>
        <option value="settings">Settings</option>
        <option aside={<Icon color={Theme.palette.spaceGrey} name="home" />} value="user">
          User
        </option>
      </Select>,
    );

    expect(render.dive()).toMatchSnapshot();
  });

  it('should render without a problem when dropdown is open', () => {
    const render = mountWithTheme(
      <Select id="select">
        <option value="home">Home</option>
        <option value="calendar">Calendar</option>
        <option value="settings">Settings</option>
        <option aside={<Icon color={Theme.palette.spaceGrey} name="home" />} value="user">
          User
        </option>
      </Select>,
    );
    render.find('#select div[onClick]').simulate('click');
    expect(render).toMatchSnapshot();
  });

  it('should render without a problem when settings is preselected and dropdown is showed', () => {
    const render = shallowWithTheme(
      <Select id="select" show selectedValue="settings">
        <option value="home">Home</option>
        <option value="calendar">Calendar</option>
        <option value="settings">Settings</option>
        <option aside={<Icon color={Theme.palette.spaceGrey} name="home" />} value="user">
          User
        </option>
      </Select>,
    );
    expect(render.dive()).toMatchSnapshot();
  });

  it('should render without a problem when props is set', () => {
    const render = shallowWithTheme(
      <Select id="select" show={false} selectedValue="settings">
        <option value="home">Home</option>
        <option value="calendar">Calendar</option>
        <option value="settings">Settings</option>
        <option aside={<Icon color={Theme.palette.spaceGrey} name="home" />} value="user">
          User
        </option>
      </Select>,
    );

    render.setProps({ selectedValue: 'calendar' });
    render.setProps({ show: true });
    expect(render.dive()).toMatchSnapshot();
  });

  it('should render without a problem when props selectedValue is unset', () => {
    const render = shallowWithTheme(
      <Select id="select" selectedValue="settings">
        <option value="home">Home</option>
        <option value="calendar">Calendar</option>
        <option value="settings">Settings</option>
        <option aside={<Icon color={Theme.palette.spaceGrey} name="home" />} value="user">
          User
        </option>
      </Select>,
    );

    render.setProps({ selectedValue: null });
    expect(render.dive()).toMatchSnapshot();
  });

  it('should render without a problem when placeholder is set', () => {
    const render = shallowWithTheme(
      <Select placeholder="Choose your menu">
        <option value="home">Home</option>
        <option value="calendar">Calendar</option>
        <option value="settings">Settings</option>
        <option aside={<Icon color={Theme.palette.spaceGrey} name="home" />} value="user">
          User
        </option>
      </Select>,
    );
    expect(render.dive()).toMatchSnapshot();
  });

  it('should render without a problem when the dropdown is open and get closed', () => {
    const render = mountWithTheme(
      <div>
        <div id="hide">Hide dropdown</div>
        <Select id="select">
          <option value="home">Home</option>
          <option value="calendar">Calendar</option>
          <option value="settings">Settings</option>
          <option aside={<Icon color={Theme.palette.spaceGrey} name="home" />} value="user">
            User
          </option>
        </Select>
        ,
      </div>,
    );
    render.find('#select div[onClick]').simulate('click');
    render.find('#select div[onBlur]').simulate('focus');
    render.find('#select div[onBlur]').simulate('blur');
    expect(render).toMatchSnapshot();
  });
});
