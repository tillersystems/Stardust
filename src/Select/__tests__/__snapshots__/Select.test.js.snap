// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render without a problem 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  text-align: left;
  padding: 0.8rem 1.2rem;
  height: 100%;
  width: 100%;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  background-color: hsl(0,100%,100%);
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.08);
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c0::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c0 > select {
  display: none;
}

<div
  className="c0"
  onBlur={[Function]}
  onClick={[Function]}
  tabIndex="0"
>
  Home
  <PoseGroup
    flipMove={true}
  />
</div>
`;

exports[`<Select /> should render without a problem when dropdown is open 1`] = `
.c1 {
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  position: absolute;
  list-style: none;
  top: calc(100% + 0.4rem);
  background-color: hsl(0,100%,100%);
  width: 100%;
  left: 0;
  max-height: 160px;
  overflow-y: auto;
  -webkit-transform-origin: 50% 0%;
  -ms-transform-origin: 50% 0%;
  transform-origin: 50% 0%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  text-align: left;
  padding: 0.8rem 1.2rem;
  height: 100%;
  width: 100%;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  background-color: hsl(0,100%,100%);
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.08);
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c0::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c0 > select {
  display: none;
}

.c2 {
  padding: 0.8rem 1.6rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0 1px 0 0 hsl(207,22%,90%);
}

.c2:last-of-type {
  box-shadow: none;
}

.c3 {
  padding-right: 0.8rem;
}

.c4 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

<Select
  id="select"
  onClick={null}
  placeholder={null}
  prefix={null}
  selectedValue={null}
  show={false}
>
  <styled.div
    onBlur={[Function]}
    onClick={[Function]}
    tabIndex="0"
  >
    <div
      className="c0"
      onBlur={[Function]}
      onClick={[Function]}
      tabIndex="0"
    >
      Home
      <PoseGroup
        flipMove={true}
      >
        <Transition
          enterAfterExit={false}
          enterPose="enter"
          exitPose="exit"
          flipMove={true}
          preEnterPose="exit"
        >
          <Component
            _pose="enter"
            initialPose="exit"
            key="DropdownAnimation"
          >
            <PoseElement
              _pose="enter"
              elementType={[Function]}
              initialPose="exit"
              poseConfig={
                Object {
                  "enter": Object {
                    "opacity": 1,
                    "scaleY": 1,
                    "transition": Object {
                      "default": Object {
                        "duration": 150,
                      },
                      "scaleY": Object {
                        "damping": 30,
                        "stiffness": 900,
                        "type": "spring",
                      },
                    },
                  },
                  "exit": Object {
                    "opacity": 0,
                    "scaleY": 0,
                    "transition": Object {
                      "duration": 150,
                    },
                  },
                }
              }
            >
              <styled.ul
                hostRef={[Function]}
                innerRef={[Function]}
              >
                <ul
                  className="c1"
                >
                  <Option
                    aside={null}
                    key=".0"
                    onClick={[Function]}
                    value="home"
                  >
                    <styled.li
                      onClick={[Function]}
                      value="home"
                    >
                      <li
                        className="c2"
                        onClick={[Function]}
                        value="home"
                      >
                        Home
                      </li>
                    </styled.li>
                  </Option>
                  <Option
                    aside={null}
                    key=".1"
                    onClick={[Function]}
                    value="calendar"
                  >
                    <styled.li
                      onClick={[Function]}
                      value="calendar"
                    >
                      <li
                        className="c2"
                        onClick={[Function]}
                        value="calendar"
                      >
                        Calendar
                      </li>
                    </styled.li>
                  </Option>
                  <Option
                    aside={null}
                    key=".2"
                    onClick={[Function]}
                    value="settings"
                  >
                    <styled.li
                      onClick={[Function]}
                      value="settings"
                    >
                      <li
                        className="c2"
                        onClick={[Function]}
                        value="settings"
                      >
                        Settings
                      </li>
                    </styled.li>
                  </Option>
                  <Option
                    aside={
                      <Styled(Icon)
                        color="hsl(207,13%,45%)"
                        name="home"
                      />
                    }
                    key=".3"
                    onClick={[Function]}
                    value="user"
                  >
                    <styled.li
                      onClick={[Function]}
                      value="user"
                    >
                      <li
                        className="c2"
                        onClick={[Function]}
                        value="user"
                      >
                        <styled.span>
                          <span
                            className="c3"
                          >
                            <Styled(Icon)
                              color="hsl(207,13%,45%)"
                              name="home"
                            >
                              <Icon
                                className=""
                                color="hsl(207,13%,45%)"
                                height="20"
                                name="home"
                                spin={false}
                                width="20"
                              >
                                <styled.div
                                  height="20"
                                  spin={false}
                                  width="20"
                                >
                                  <div
                                    className="c4"
                                    height="20"
                                    width="20"
                                  >
                                    <svg
                                      height="20"
                                      style={
                                        Object {
                                          "display": "block",
                                        }
                                      }
                                      viewBox="0 0 512 512"
                                      width="20"
                                      xmlns="http://www.w3.org/2000/svg"
                                    >
                                      <path
                                        d="M341.333 448h64c11.782 0 21.334-9.551 21.334-21.333V202.434L256 69.693 85.333 202.433v224.234c0 11.782 9.552 21.333 21.334 21.333h64V256c0-11.782 9.55-21.333 21.333-21.333h128c11.782 0 21.333 9.55 21.333 21.333v192zm-42.666 0V277.333h-85.334V448h85.334zM50.903 175.16l192-149.333a21.333 21.333 0 0 1 26.194 0l192 149.333a21.333 21.333 0 0 1 8.236 16.84v234.667c0 35.346-28.653 64-64 64H106.667c-35.347 0-64-28.654-64-64V192c0-6.583 3.04-12.798 8.236-16.84z"
                                        fill="hsl(207,13%,45%)"
                                        fillRule="evenodd"
                                      />
                                    </svg>
                                  </div>
                                </styled.div>
                              </Icon>
                            </Styled(Icon)>
                          </span>
                        </styled.span>
                        User
                      </li>
                    </styled.li>
                  </Option>
                </ul>
              </styled.ul>
            </PoseElement>
          </Component>
        </Transition>
      </PoseGroup>
    </div>
  </styled.div>
</Select>
`;

exports[`<Select /> should render without a problem when placeholder is set 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  text-align: left;
  padding: 0.8rem 1.2rem;
  height: 100%;
  width: 100%;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  background-color: hsl(0,100%,100%);
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.08);
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c0::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c0 > select {
  display: none;
}

<div
  className="c0"
  onBlur={[Function]}
  onClick={[Function]}
  tabIndex="0"
>
  <styled.span>
    Choose your menu
  </styled.span>
  <PoseGroup
    flipMove={true}
  />
</div>
`;

exports[`<Select /> should render without a problem when props is set 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  text-align: left;
  padding: 0.8rem 1.2rem;
  height: 100%;
  width: 100%;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  background-color: hsl(0,100%,100%);
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.08);
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c0::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c0 > select {
  display: none;
}

<div
  className="c0"
  onBlur={[Function]}
  onClick={[Function]}
  tabIndex="0"
>
  Calendar
  <PoseGroup
    flipMove={true}
  >
    <Component
      key="DropdownAnimation"
    >
      <Option
        aside={null}
        key=".0"
        onClick={[Function]}
        value="home"
      >
        Home
      </Option>
      <Option
        aside={null}
        key=".1"
        onClick={[Function]}
        value="calendar"
      >
        Calendar
      </Option>
      <Option
        aside={null}
        key=".2"
        onClick={[Function]}
        value="settings"
      >
        Settings
      </Option>
      <Option
        aside={
          <Styled(Icon)
            color="hsl(207,13%,45%)"
            name="home"
          />
        }
        key=".3"
        onClick={[Function]}
        value="user"
      >
        User
      </Option>
    </Component>
  </PoseGroup>
</div>
`;

exports[`<Select /> should render without a problem when props selectedValue is unset 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  text-align: left;
  padding: 0.8rem 1.2rem;
  height: 100%;
  width: 100%;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  background-color: hsl(0,100%,100%);
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.08);
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c0::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c0 > select {
  display: none;
}

<div
  className="c0"
  onBlur={[Function]}
  onClick={[Function]}
  tabIndex="0"
>
  Home
  <PoseGroup
    flipMove={true}
  />
</div>
`;

exports[`<Select /> should render without a problem when settings is preselected and dropdown is showed 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  text-align: left;
  padding: 0.8rem 1.2rem;
  height: 100%;
  width: 100%;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  background-color: hsl(0,100%,100%);
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.08);
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c0::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c0 > select {
  display: none;
}

<div
  className="c0"
  onBlur={[Function]}
  onClick={[Function]}
  tabIndex="0"
>
  Settings
  <PoseGroup
    flipMove={true}
  />
</div>
`;

exports[`<Select /> should render without a problem when the dropdown is open and get closed 1`] = `
.c1 {
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  position: absolute;
  list-style: none;
  top: calc(100% + 0.4rem);
  background-color: hsl(0,100%,100%);
  width: 100%;
  left: 0;
  max-height: 160px;
  overflow-y: auto;
  -webkit-transform-origin: 50% 0%;
  -ms-transform-origin: 50% 0%;
  transform-origin: 50% 0%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  text-align: left;
  padding: 0.8rem 1.2rem;
  height: 100%;
  width: 100%;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 4px;
  background-color: hsl(0,100%,100%);
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.08);
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  right: 1.4rem;
  top: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-right-width: 3px;
  border-left-width: 3px;
  border-bottom: 3px solid hsl(207,13%,45%);
}

.c0::before {
  content: '';
  position: absolute;
  right: 1.4rem;
  bottom: 37%;
  width: 0;
  height: 0;
  border: 0 solid transparent;
  border-left-width: 3px;
  border-right-width: 3px;
  border-top: 3px solid hsl(207,13%,45%);
}

.c0 > select {
  display: none;
}

.c2 {
  padding: 0.8rem 1.6rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0 1px 0 0 hsl(207,22%,90%);
}

.c2:last-of-type {
  box-shadow: none;
}

.c3 {
  padding-right: 0.8rem;
}

.c4 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

<div>
  <div
    id="hide"
  >
    Hide dropdown
  </div>
  <Select
    id="select"
    onClick={null}
    placeholder={null}
    prefix={null}
    selectedValue={null}
    show={false}
  >
    <styled.div
      onBlur={[Function]}
      onClick={[Function]}
      tabIndex="0"
    >
      <div
        className="c0"
        onBlur={[Function]}
        onClick={[Function]}
        tabIndex="0"
      >
        Home
        <PoseGroup
          flipMove={true}
        >
          <Transition
            enterAfterExit={false}
            enterPose="enter"
            exitPose="exit"
            flipMove={true}
            preEnterPose="exit"
          >
            <Component
              _pose="exit"
              initialPose="exit"
              key="DropdownAnimation"
              onPoseComplete={[Function]}
              popFromFlow={true}
            >
              <PoseElement
                _pose="exit"
                elementType={[Function]}
                initialPose="exit"
                onPoseComplete={[Function]}
                popFromFlow={true}
                poseConfig={
                  Object {
                    "enter": Object {
                      "opacity": 1,
                      "scaleY": 1,
                      "transition": Object {
                        "default": Object {
                          "duration": 150,
                        },
                        "scaleY": Object {
                          "damping": 30,
                          "stiffness": 900,
                          "type": "spring",
                        },
                      },
                    },
                    "exit": Object {
                      "opacity": 0,
                      "scaleY": 0,
                      "transition": Object {
                        "duration": 150,
                      },
                    },
                  }
                }
              >
                <styled.ul
                  hostRef={[Function]}
                  innerRef={[Function]}
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "position": "absolute",
                      "top": 0,
                      "width": 0,
                    }
                  }
                >
                  <ul
                    className="c1"
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "position": "absolute",
                        "top": 0,
                        "width": 0,
                      }
                    }
                  >
                    <Option
                      aside={null}
                      key=".0"
                      onClick={[Function]}
                      value="home"
                    >
                      <styled.li
                        onClick={[Function]}
                        value="home"
                      >
                        <li
                          className="c2"
                          onClick={[Function]}
                          value="home"
                        >
                          Home
                        </li>
                      </styled.li>
                    </Option>
                    <Option
                      aside={null}
                      key=".1"
                      onClick={[Function]}
                      value="calendar"
                    >
                      <styled.li
                        onClick={[Function]}
                        value="calendar"
                      >
                        <li
                          className="c2"
                          onClick={[Function]}
                          value="calendar"
                        >
                          Calendar
                        </li>
                      </styled.li>
                    </Option>
                    <Option
                      aside={null}
                      key=".2"
                      onClick={[Function]}
                      value="settings"
                    >
                      <styled.li
                        onClick={[Function]}
                        value="settings"
                      >
                        <li
                          className="c2"
                          onClick={[Function]}
                          value="settings"
                        >
                          Settings
                        </li>
                      </styled.li>
                    </Option>
                    <Option
                      aside={
                        <Styled(Icon)
                          color="hsl(207,13%,45%)"
                          name="home"
                        />
                      }
                      key=".3"
                      onClick={[Function]}
                      value="user"
                    >
                      <styled.li
                        onClick={[Function]}
                        value="user"
                      >
                        <li
                          className="c2"
                          onClick={[Function]}
                          value="user"
                        >
                          <styled.span>
                            <span
                              className="c3"
                            >
                              <Styled(Icon)
                                color="hsl(207,13%,45%)"
                                name="home"
                              >
                                <Icon
                                  className=""
                                  color="hsl(207,13%,45%)"
                                  height="20"
                                  name="home"
                                  spin={false}
                                  width="20"
                                >
                                  <styled.div
                                    height="20"
                                    spin={false}
                                    width="20"
                                  >
                                    <div
                                      className="c4"
                                      height="20"
                                      width="20"
                                    >
                                      <svg
                                        height="20"
                                        style={
                                          Object {
                                            "display": "block",
                                          }
                                        }
                                        viewBox="0 0 512 512"
                                        width="20"
                                        xmlns="http://www.w3.org/2000/svg"
                                      >
                                        <path
                                          d="M341.333 448h64c11.782 0 21.334-9.551 21.334-21.333V202.434L256 69.693 85.333 202.433v224.234c0 11.782 9.552 21.333 21.334 21.333h64V256c0-11.782 9.55-21.333 21.333-21.333h128c11.782 0 21.333 9.55 21.333 21.333v192zm-42.666 0V277.333h-85.334V448h85.334zM50.903 175.16l192-149.333a21.333 21.333 0 0 1 26.194 0l192 149.333a21.333 21.333 0 0 1 8.236 16.84v234.667c0 35.346-28.653 64-64 64H106.667c-35.347 0-64-28.654-64-64V192c0-6.583 3.04-12.798 8.236-16.84z"
                                          fill="hsl(207,13%,45%)"
                                          fillRule="evenodd"
                                        />
                                      </svg>
                                    </div>
                                  </styled.div>
                                </Icon>
                              </Styled(Icon)>
                            </span>
                          </styled.span>
                          User
                        </li>
                      </styled.li>
                    </Option>
                  </ul>
                </styled.ul>
              </PoseElement>
            </Component>
          </Transition>
        </PoseGroup>
      </div>
    </styled.div>
  </Select>
  ,
</div>
`;
